File: tests/fixtures/polyglot/ts/billing/tax.ts
import { LedgerEntry } from "./ledger";

/** Rates applied to an invoice line. */
export const STANDARD_RATE = 0.2;

export class TaxCalculator {
  constructor(private rate: number = STANDARD_RATE) {}

  /** Compute the tax owed on an amount. */
  taxFor(amount: number): number {
    return round2(amount * this.rate);
  }

  /** Total including tax. */
  totalWithTax(amount: number): number {
    return round2(amount + this.taxFor(amount));
  }
}

export function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

export function invoiceTotal(entries: LedgerEntry[]): number {
  const calc = new TaxCalculator();
  return entries.reduce((sum, e) => sum + calc.totalWithTax(e.amount), 0);
}

File: tests/fixtures/polyglot/rust/billing/tax.rs
//! Tax computation for invoices.

/// Rate applied to a standard-rated invoice line.
pub const STANDARD_RATE: f64 = 0.2;

/// A single billable invoice line.
pub struct Invoice {
    pub id: String,
    pub amount: f64,
}

/// Computes tax owed on an amount.
pub struct TaxCalculator {
    rate: f64,
}

impl TaxCalculator {
    /// Build a calculator using the standard rate.
    pub fn new() -> Self {
        TaxCalculator { rate: STANDARD_RATE }
    }

    /// Tax owed on `amount`.
    pub fn tax_for(&self, amount: f64) -> f64 {
        round2(amount * self.rate)
    }

    /// Amount including tax.
    pub fn total_with_tax(&self, amount: f64) -> f64 {
        round2(amount + self.tax_for(amount))
    }
}

/// Round to two decimal places.
pub fn round2(value: f64) -> f64 {
    (value * 100.0).round() / 100.0
}

/// Sum the taxed totals of every invoice.
pub fn invoice_total(invoices: &[Invoice]) -> f64 {
    let calc = TaxCalculator::new();
    let mut total = 0.0;
    for inv in invoices {
        total += calc.total_with_tax(inv.amount);
    }
    total
}
